tests/exponential.rs

```diff
@@ -0,0 +1,402 @@
+}
+
+impl SynthLanguage for Exponential {
+    type Constant = Rational;
+
+    // cvec-less domain
+    fn eval<'a, F>(&'a self, _cvec_len: usize, _get_cvec: F) -> CVec<Self>
+    where
+        F: FnMut(&'a Id) -> &'a CVec<Self>,
+    {
+        vec![]
+    }
+
+    fn initialize_vars(egraph: &mut EGraph<Self, SynthAnalysis>, vars: &[String]) {
+        for v in vars {
+            let id = egraph.add(Exponential::Var(Symbol::from(v.clone())));
+
+            let l_id = egraph.add(Exponential::Log(id));
+            let el_id = egraph.add(Exponential::Exp(l_id));
+
+            let e_id = egraph.add(Exponential::Exp(id));
+            let le_id = egraph.add(Exponential::Log(e_id));
+
+            egraph.union(id, el_id);
+            egraph.union(id, le_id);
+            egraph.rebuild();
+        }
+    }
+
+    fn to_var(&self) -> Option<Symbol> {
+        if let Exponential::Var(sym) = self {
+            Some(*sym)
+        } else {
+            None
+        }
+    }
+
+    fn mk_var(sym: Symbol) -> Self {
+        Exponential::Var(sym)
+    }
+
+    fn is_constant(&self) -> bool {
+        matches!(self, Exponential::Num(_))
+    }
+
+    fn mk_constant(c: Self::Constant, _egraph: &mut EGraph<Self, SynthAnalysis>) -> Self {
+        Exponential::Num(c)
+    }
+
+    // no validation possible
+    fn validate(_lhs: &Pattern<Self>, _rhs: &Pattern<Self>) -> ValidationResult {
+        ValidationResult::Valid
+    }
+
+    fn is_rule_lifting() -> bool {
+        true
+    }
+
+    fn get_lifting_rules() -> enumo::Ruleset<Self> {
+        enumo::Ruleset::from_str_vec(&[
+            // definitions (denote)
+            "(pow ?a ?b) ==> (exp (* ?b (log ?a)))",
+            "(sqrt ?a) ==> (pow ?a 1/2)",
+            "(cbrt ?a) ==> (pow ?a 1/3)",
+            // definitions (simplify)
+            "(exp (* ?b (log ?a))) ==> (pow ?a ?b)",
+            "(pow ?a 1/2) ==> (sqrt ?a)",
+            "(pow ?a 1/3) ==> (cbrt ?a)",
+        ])
+    }
+
+    fn is_allowed_op(&self) -> bool {
+        true
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use ruler::enumo;
+
+    type Workload = enumo::Workload;
+    type Ruleset = enumo::Ruleset<Exponential>;
+    type Filter = enumo::Filter;
+
+    macro_rules! str_vec {
+        ($($x:expr),*) => (vec![$($x.to_string()),*]);
+    }
+
+    fn starting_exponential_rules() -> Ruleset {
+        Ruleset::from_str_vec(&[
+            // exponential properties (expand)
+            "(exp (+ ?a ?b)) ==> (* (exp ?a) (exp ?b))",
+            "(exp (~ ?a)) ==> (/ 1 (exp ?a))",
+            // exponential properties (simplify)
+            "(* (exp ?a) (exp ?b)) ==> (exp (+ ?a ?b))",
+            "(/ 1 (exp ?a)) ==> (exp (~ ?a))",
+            "(exp 0) ==> 1",
+            // inverse properties
+            "(log (exp ?a)) ==> ?a",
+            "(exp (log ?a)) ==> ?a",
+        ])
+    }
+
+    // TODO: actually derive these rules
+    fn rational_rules() -> Ruleset {
+        Ruleset::from_str_vec(&[
```

Comment:
what are these? could you use baseline/rational.rules? or add a new file to baseline/ ?

Reply:
Oh yep. Forgot about that.